nano: update to 2.8.4

hnyman opened the pull request:
* update nano to 2.8.1
* remove deprecated glibtest configure argument
* fix whitespace in configure arguments

I am not sure if this is ready for merge yet, as I noticed that the size of the binary file has doubled from 2.7.5. Might be due to the gnulib usage since 2.8.0. Some additional tweaking might be needed to keep the size down.

Maintainer: @jp-bennett 
Run tested: mvebu WRT3200ACM with LEDE master

https://www.nano-editor.org/news.php

> 2017 April 12
> 
> GNU nano 2.8.1 "Ellert" fixes build failures on MacOS and on musl, fixes scrolling problems in softwrap mode when double-width characters on row boundaries are involved, shows double-width characters as ">" and "<" when split across two rows, moves the cursor more predictably (at the cost of sometimes putting it on the second "half" of a character), avoids creating lines that consist of only blanks when using autoindent, makes ^Home and ^End go to the start and end of the file (on terminals that support those keystrokes), places the cursor better when linting, lets the linter ask only once whether to open an included file, and adds bindings for ^Up and ^Down in the file browser.  Don't sit on your hands.
> 
> 2017 March 31
> 
> GNU nano 2.8.0 "Axat" makes it easier to move around in softwrapped lines: the Up and Down keys now step from visual row to visual row instead of jumping between logical lines, and the Home and End keys now move to the start and end of a row, and only when already there, then to the start and end of the logical line. Furthermore, the screen can now scroll per row instead of always per logical line.  On an entirely different front: nano now makes use of gnulib, to make it build on more platforms.  In short: there were many internal changes, not many user-visible ones (apart form the new softwrap navigation).  The conversion to gnulib was done by Mike Frysinger, the softwrap overhaul by David Ramsey.

jp-bennett commented:
I'd like to look into the binary size as well before we pull this.  Thanks for working on keeping this up to date over the last few months while I've been slammed with other work.

diizzyy commented:
We probably want unicode unless it's dramatically increases binary size?

diizzyy commented:
...without unicode support.
```
        --enable-threads=pth \
        --disable-libmagic \
```
~64kbyte ipk 
Binary:  122929 bytes

Adding
```
TARGET_CFLAGS += -ffunction-sections -fdata-sections
TARGET_LDFLAGS += -Wl,--gc-sections
```
~63kbyte ipk
Binary: 122929 bytes

diizzyy commented:
I'm also getting 4 bytes errors when opening files.

hnyman commented:
> "Unexpected large character size: 4 bytes -- please report a bug"

I got now the same error.
The error seems to materialize just by launching nano.
Possible due to our configure options that confuse some logic.

hnyman commented:
I debugged this "4 byte" warning a bit further. It is likely just a cosmetic warning caused by the combination of nano 2.8.1 and musl.

nano 2.8.1 has a new warning about possible inconsistency (no UTF8 support active in nano while according to locale the max. char length is more than 1). Musl libc sets MB_CUR_MAX to 4 unless the locale is explicitly set to C, when the value is 1. And that seems to trigger the new warning in nano as 4>1.

Relevant commits:
* http://git.savannah.gnu.org/cgit/nano.git/commit/?id=aedc3ddd49b50fcee73ad76e193ab0a2c68f4adf
* http://git.musl-libc.org/cgit/musl/commit/?id=1507ebf837334e9e07cfab1ca1c2e88449069a80
* http://git.musl-libc.org/cgit/musl/commit/?id=f22a9edaf8a6f2ca1d314d18b3785558279a5c03

hnyman commented:
I filed a bug/question about it in upstream:
https://savannah.gnu.org/bugs/index.php?50897

Compiling nano with UTF8 support makes the warning go away but increases its size by some 4%.

hnyman commented:
Updated with upstream patch and rebased.

Upstream nano developer agreed that the character size warning is cosmetics and he patched it away with http://git.savannah.gnu.org/cgit/nano.git/commit/?id=a8a60b29e679c33c8abbeb36875662a3dd26e4cb

> Yes, the warning can be seen as cosmetics -- no harm will happen.
> 
> In fact, all I wanted to know is whether there are systems that have an MB_CUR_MAX bigger than 6 when using UTF-8. I didn't think that anyone would compile with --disable-utf8. So now, for that case, the warning does not get included.
> 
> Fixed in git, a8a60b29.

hnyman commented:
Upstream has released 2.8.2 that includes the fix

hnyman commented:
Updated for 2.8.2

hnyman commented:
I don't think that there is much that can be done for making the size smaller. I guess that we might just have to bite the bullet and accept the size growth.

diizzyy commented:
What's the size without threads and libmagic?

hnyman commented:
Rebased and updated for 2.8.4

> What's the size without threads and libmagic?

Based on docs, "enable-tiny" is already includes also "disable-libmagic"...
What about threads? I do not see a config option for that.

@jp-bennett 
I propose that we merge this in. The binary size will grow a bit, but as the upstream has changed the composition of built-in libraries, there is not much we can do. Minor tweaking could maybe be done with gcc options etc., but that seems cumbersome.

diizzyy commented:
```
nano-2.8.4 % ./configure --help
```
```
Optional Features:
....
  --enable-threads={posix|solaris|pth|windows}
                          specify multithreading API
....
```

hnyman commented:
Funny. I checked just the configure.ac and found nothing about threads. :-(

hnyman commented:
The threads=pth change makes no significant change in either .ipk or executable size.

Normal:
ipk: 64178
nano: 122929

threads=pth
ipk: 63979
nano: 122929

D-Albers commented:
In which case does the ipk size really matter?

I would say, that only the executable size is important and the value is equal in both cases.

hnyman commented:
I merge this.
Let's look for the size tweaks later.
